python/lib/core/dmod/core/common/protocols.py

```diff
@@ -56,6 +58,209 @@ class DescribableProtocol(typing.Protocol):
+        timeout seconds. If the call hasn’t completed in timeout seconds, then a concurrent.futures.TimeoutError
+        will be raised. timeout can be an int or float. If timeout is not specified or None, there is no limit to the
+        wait time.
+
+        If the future is cancelled before completing then CancelledError will be raised.
+
+        If the call raised, this method will raise the same exception.
+
+        Args:
+            timeout: The number of seconds to wait for a result
+
+        Returns:
+            The result of the call
+        """
+
+    def exception(self, timeout: typing.Union[int, float] = None) -> typing.Optional[BaseException]:
+        """
+        Return the exception raised by the call. If the call hasn’t yet completed then this method will wait up to
+        timeout seconds. If the call hasn’t completed in timeout seconds, then a TimeoutError will
+        be raised. timeout can be an int or float. If timeout is not specified or None, there is no limit to the wait
+        time.
+
+        If the future is cancelled before completing then CancelledError will be raised.
+
+        If the call completed without raising, None is returned.
+
+        Args:
+            timeout: The number of seconds to wait
+
+        Returns:
+            An exception that was raised by the call, if one was raised, else None
+        """
+
+    def add_done_callback(self, fn: typing.Callable[[Self], typing.Any]):
+        """
+        Attaches the callable fn to the future. fn will be called, with the future as its only argument, when the
+        job is cancelled or finishes running.
+
+        Added callables are called in the order that they were added and are always called in a thread belonging to the
+        process that added them. If the callable raises an Exception subclass, it will be logged and ignored. If the
+        callable raises a BaseException subclass, the behavior is undefined.
+
+        If the job has already completed or been cancelled, fn will be called immediately.
+
+        Args:
+            fn: The function to call when the job is done
+        """
+
+
+@typing.runtime_checkable
+class JobLauncherProtocol(typing.Protocol):
+    """
+    Represents an object that may launch actions outside of the current thread and process
+    """
+    def submit(self, fn: typing.Callable, *args, **kwargs) -> JobResultProtocol:
+        """
+        Schedules the callable, fn, to be executed as fn(*args **kwargs) and returns a JobResultProtocol object
+        representing the execution of the callable.
+
+        Args:
+            fn: The function to call
+            *args: positional arguments for the function
+            **kwargs: Keyword arguments for the function
+
+        Returns:
+            A JobResultProtocol object representing the result of the function
+        """
+
+    def map(
+        self,
+        *iterables,
+        timeout: typing.Union[int, float] = None,
```

Comment:
Ask Guido - the class is just a typed interface for the `Executor`s. That's what the `Executor` defines, so that's what's there. Aside from the `func` parameter. That's just a bug.